### PR TITLE
[MM-45444] Denormalize Reactions to add ChannelId for top reactions insights query

### PR DESCRIPTION
**Pull request by plant99**
#### Summary

Top Reactions for team is slow and taking 9s in community.mattermost.com. This change adds `ChannelId` to `Reactions` to avoid joins with `Posts` table.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-45444
#### Release Note

```release-note
Reduce API times for `My Team: Top Reactions` insights.
Add notes from https://github.com/mattermost/mattermost-server/pull/20572#issuecomment-1176238813 to upgrade docs.
```

### CODE REVIEW

**Review by mkraft — COMMENTED** [review]
What about the `create index reactions_deleteat_createat on reactions(deleteat, createat);`? Did you determine it wasn't beneficial?

File: api4/reaction_test.go [hunk 1]
```diff
@@ -326,6 +326,7 @@ func TestDeleteReaction(t *testing.T) {
 
 	t.Run("delete-reaction-when-post-has-multiple-reactions", func(t *testing.T) {
 		th.App.SaveReactionForPost(th.Context, r1)
+		r1.ChannelId = ""
```
[the remark this answers is not in the document]
Comment (plant99) [reply api4/reaction_test.go:329]: So `Save` returns `Reaction` with `ChannelId` and some form of `GetBulkReactions` uses a different select query. I'll include ChannelId in the query and remove Lines 329, 561.

File: db/migrations/postgres/000089_add-channelid-to-reaction.up.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE reactions ADD COLUMN IF NOT EXISTS channelid varchar(26);
+UPDATE reactions SET channelid = (select channelid from posts where posts.id = reactions.postid);
+ALTER TABLE reactions ALTER column channelid set not null;
```
[the remark this answers is not in the document]
Comment (plant99) [reply db/migrations/postgres/000089_add-channelid-to-reaction.up.sql:3]: I'll set it to `not null` in MySQL as well. That'd be safe.

**Comment by agnivade** [comment]
1. Capture the time taken for the SQL queries for both MySQL and Postgres, along with the number of rows and DB specs.

2.
```sql
ALTER TABLE reactions ADD COLUMN IF NOT EXISTS channelid varchar(26) NOT NULL;
UPDATE reactions SET channelid = (select channelid from posts where posts.id = reactions.postid);
CREATE INDEX IF NOT EXISTS idx_reactions_channel_id on reactions (channelid);
```

Do the same for MySQL as well.

**Comment by agnivade** [comment]
3. Ask the SRE team to merge it at their convenience.

File: db/migrations/mysql/000089_add-channelid-to-reaction.up.sql [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    ),
+    'ALTER TABLE Reactions ADD COLUMN ChannelId varchar(26);',
+    'SELECT 1;'
+));
+
+PREPARE addColumnIfNotExists FROM @preparedStatement;
+EXECUTE addColumnIfNotExists;
+DEALLOCATE PREPARE addColumnIfNotExists;
+
+UPDATE Reactions SET ChannelId = (select ChannelId from Posts where Posts.Id = Reactions.PostId);
```
[the remark this answers is not in the document]
Comment (plant99) [reply db/migrations/mysql/000089_add-channelid-to-reaction.up.sql:16]: I misunderstood, will add it in the next commit.

File: db/migrations/postgres/000089_add-channelid-to-reaction.down.sql [hunk 4]
```diff
@@ -0,0 +1 @@
+ALTER TABLE reactions drop column channelid;
```
Comment (agnivade) [inline db/migrations/postgres/000089_add-channelid-to-reaction.down.sql:1]: This needs to be idempotent: "DROP COLUMN IF EXISTS"

File: store/sqlstore/reaction_store.go [hunk 5]
```diff
@@ -26,7 +26,14 @@ func (s *SqlReactionStore) Save(reaction *model.Reaction) (*model.Reaction, erro
 	if err := reaction.IsValid(); err != nil {
 		return nil, err
 	}
-
+	if reaction.ChannelId == "" {
+		// get channelId before beginning transaction, if not already populated
+		chIds, err := s.Post().GetPostsByIds([]string{reaction.PostId})
```
Comment (agnivade) [inline store/sqlstore/reaction_store.go:31]: This does lots of extra things. I'd suggest to begin the transaction earlier, and just write a raw SQL query `select channelId from posts where id=?` to directly get the channel id.

**Comment by plant99** [comment]
### Stats for PostgreSQL
```sql
-- timing and db stats
/*

db.r5.2xlarge - aurora postgresql

2.5 million reactions
12 million posts
*/

ALTER TABLE reactions ADD COLUMN IF NOT EXISTS channelid varchar(26) NOT NULL DEFAULT ''; -- 6.978 ms
UPDATE reactions SET channelid = (select channelid from posts where posts.id = reactions.postid); -- 68.69 seconds
CREATE INDEX IF NOT EXISTS idx_reactions_channel_id on reactions (channelid); -- 3 seconds
```

Will update for MySQL tomorrow. Meanwhile, please re-review the PR @agnivade

**Review by agnivade — APPROVED** [review]

**Comment by mkraft** [comment]
@plant99 Should we remove [this join on Posts](https://github.com/mattermost/mattermost-server/pull/20572/files#diff-3325531dd7329db0673a7ac15ebc721750566f5227f44b68c28dfbc434034672L309), too? Or do you think it would be best to do that for the next release?

**Comment by plant99** [comment]
> @plant99 Should we remove [this join on Posts](https://github.com/mattermost/mattermost-server/blob/cea6662bc4e313456ede475ac3a938f0b33efbff/store/sqlstore/reaction_store.go#L319), too? Or do you think it would be best to do that for the next release?

You're right about removing that join, but considering the time constraint, I didn't want to touch any more parts of code than what's necessary.

It seems like a two-liner though + a few other test changes, so a fix would be easy. We'll create a ticket for 7.2, if that's fine?

**Comment by mkraft** [comment]
> > @plant99 Should we remove [this join on Posts](https://github.com/mattermost/mattermost-server/blob/cea6662bc4e313456ede475ac3a938f0b33efbff/store/sqlstore/reaction_store.go#L319), too? Or do you think it would be best to do that for the next release?
> 
> You're right about removing that join, but considering the time constraint, I didn't want to touch any more parts of code than what's necessary.
> 
> It seems like a two-liner though + a few other test changes, so a fix would be easy. We'll create a ticket for 7.2, if that's fine?

I think that's a great idea.

**Comment by plant99** [comment]
@cwarnermm @mkraft @agnivade  could you please review these upgrade notes?

```
New schema changes were introduced in the form of a new column, and its index. The following summarizes the test results measuring how long it took for the database queries to run with these schema changes:

MySQL 12M Posts, 2.5M Reactions - ~1min 34s (Instance: PC with 8 cores, 16GB RAM)
Postgres 12M Posts, 2.5M Reactions - ~1min 18s (Instance: db.r5.2xlarge)

For customers wanting an upgrade with substantially reduced downtime(~1/5th), they are encouraged to run the following queries before upgrading. This is fully backwards compatible but will obtain a lock on Reactions table, and users' reactions in the interval wouldn't be reflected in the DB.

For MySQL:
ALTER TABLE Reactions ADD COLUMN ChannelId varchar(26) NOT NULL DEFAULT "";
UPDATE Reactions SET ChannelId = (select ChannelId from Posts where Posts.Id = Reactions.PostId) WHERE ChannelId="";
CREATE INDEX idx_reactions_channel_id ON Reactions(ChannelId) LOCK=NONE;

For Postgres:
ALTER TABLE reactions ADD COLUMN IF NOT EXISTS channelid varchar(26) NOT NULL DEFAULT '';
UPDATE reactions SET channelid = (select channelid from posts where posts.id = reactions.postid) WHERE channelid='';
CREATE INDEX CONCURRENTLY IF NOT EXISTS idx_reactions_channel_id on reactions (channelid);
```

Also Martin, Agniva - please review [this commit](https://github.com/mattermost/mattermost-server/pull/20572/commits/e8eaad17df44e13eee5c7ae6c4410b562df23fbd). Without that change, if the deployment had pre-migrations, it'd still take time to update all the rows. (I'm working to test this on a large PostgresDB)

File: db/migrations/mysql/000089_add-channelid-to-reaction.up.sql [hunk 6]
```diff
@@ -14,7 +14,7 @@ EXECUTE addColumnIfNotExists;
 DEALLOCATE PREPARE addColumnIfNotExists;
 
 
-UPDATE Reactions SET ChannelId = (select ChannelId from Posts where Posts.Id = Reactions.PostId);
+UPDATE Reactions SET ChannelId = (select ChannelId from Posts where Posts.Id = Reactions.PostId) WHERE ChannelId="";
```
Comment (mkraft) [inline db/migrations/mysql/000089_add-channelid-to-reaction.up.sql:17]: Good thinking.

Comment (agnivade) [reply db/migrations/mysql/000089_add-channelid-to-reaction.up.sql:17]: Indeed. Always keep in mind that queries need to be idempotent. Nice work :+1:

**Pull request merged**